lib/msf/core/post/windows/filesystem.rb

```diff
@@ -0,0 +1,342 @@
+
+
+  def set_reparse_point(handle, reparse_buffer)
+    result = session.railgun.kernel32.DeviceIoControl(
+      handle,
+      FSCTL_SET_REPARSE_POINT,
+      reparse_buffer,
+      reparse_buffer.size,
+      nil,
+      0,
+      4,
+      nil
+    )
+
+    unless result['return']
+      print_error("Error setting the reparse point. Windows Error Code: #{result['GetLastError']} - #{result['ErrorMessage']}")
+    end
+    result['return']
+  end
+
+  def delete_reparse_point(handle, reparse_buffer)
+    result = session.railgun.kernel32.DeviceIoControl(
+      handle,
+      FSCTL_DELETE_REPARSE_POINT,
+      reparse_buffer,
+      reparse_buffer.size,
+      nil,
+      0,
+      4,
+      nil
+    )
+
+    unless result['return']
+      print_error("Error deleting the reparse point. Windows Error Code: #{result['GetLastError']} - #{result['ErrorMessage']}")
+    end
+    result['return']
+  end
+
+  def open_reparse_point(path, writable)
+    result = session.railgun.kernel32.CreateFileW(
+      path,
+      "GENERIC_READ | #{writable ? 'GENERIC_WRITE' : '0'}",
+      'FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE',
+      nil,
+      'OPEN_EXISTING',
+      'FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT',
+      0
+    )
+
+    handle = result['return']
+    if handle.nil? || handle == INVALID_HANDLE_VALUE
+      print_error("Error opening #{path}. Windows Error Code: #{result['GetLastError']} - #{result['ErrorMessage']}")
+      return nil
+    end
+    vprint_good("Successfuly opened #{path}")
+    handle
+  end
+
+  def delete_mount_point(path)
+    buffer = ReparseGuidDataBuffer.new
+    buffer.reparse_tag = IO_REPARSE_TAG_MOUNT_POINT
+
+    handle = open_reparse_point(path, true);
+    return nil unless handle
+    delete_reparse_point(handle, buffer.to_binary_s)
+  end
+
+  def write_to_memory(process, str)
+    p_buffer = process.memory.allocate(str.size)
+    unless p_buffer
+      print_error("Error alocating memory for \"#{str}\": Windows Error Code: #{result['GetLastError']} - #{result['ErrorMessage']}")
+      return nil
+    end
+    unless process.memory.write(p_buffer, str) == str.size
+      print_error("Error writting \"#{str}\" to memory buffer: Windows Error Code: #{result['GetLastError']} - #{result['ErrorMessage']}")
+      return nil
+    end
+    p_buffer
+  end
+
+  def build_unicode_string(str_byte_size, p_buffer)
+    unicode_str = UnicodeString.new(
+      arch: client.native_arch == ARCH_X64 ? UnicodeString::ARCH_X64 : UnicodeString::ARCH_X86
+    )
+    unicode_str.str_length = str_byte_size - 2
+    unicode_str.maximum_length = str_byte_size
+    unicode_str.p_buffer = p_buffer
+    unicode_str
+  end
+
+  def str_to_unicode(str)
+    str.encode('UTF-16LE').force_encoding('binary') + "\x00\x00"
+  end
+
+  def setup_unicode_str_in_memory(process, str)
+    enc_str = str_to_unicode(str)
+    p_buffer = write_to_memory(process, enc_str)
+    return nil unless p_buffer
```

Comment:
Most of this logic could be replaced with the [`alloc_and_write_wstring`](https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/post/meterpreter/extensions/stdapi/railgun/util.rb#L384) railgun utility function.

Reply:
We are going to hold off on this